### ISSUE TITLE
Maybe add some mechanism to auto-merge PRs

### ISSUE THREAD

**Issue by MartinNowak** [dlang-bots/dlang-bot#11]
Estimate possible approaches and effort first.

Do we need user OAuth tokens to merge on their behalf?

Could we live w/o the merge author in git and just use the dlang-bot account?

What buttons to use? I think labels were proposed b/c they're already only editable by maintainers.

**Comment by wilzbach** [comment]
> Do we need user OAuth tokens to merge on their behalf?

Well as long as git commits aren't signed, we can create the git commit on their behalf exactly as they would do locally. If you want to do it via the GitHub API, then yes we would need an OAuth token.

We could do the merge ourselves with dlang-bot, set the correct merge author and then push it.
Though that's probably more risky and work than merging with dlang-bot.

> Could we live w/o the merge author in git and just use the dlang-bot account?

Idk, but sometimes it's quite nice to see how signed off a commit. On the other hand GitHub also saves the entire pull request history (e.g. who has set the label), so we wouldn't loose this information, but have to trust that GitHub doesn't screw up the info (unlikely).

> What buttons to use? I think labels were proposed b/c they're already only editable by maintainers.

I agree - that's the simplest :)
(though there's the gotcha that labels don't trigger notification, but dlang-bot could post a comment as "warning")

**Comment by MartinNowak** [comment]
> (though there's the gotcha that labels don't trigger notification, but dlang-bot could post a comment as "warning")

Not using notifications myself. What's the problem w/ that? You'd expect to get notified if someone turns on auto-merge? Does changing a comment trigger a new notification?
I'm actually not a friend of all the many bot comments, makes PRs sometimes very hard to read.
Label changes flow with the PR comment timeline, so that would be good enough IMO.

**Comment by wilzbach** [comment]
> What's the problem w/ that? You'd expect to get notified if someone turns on auto-merge?

Yes that's the current behavior. 

> Does changing a comment trigger a new notification?

No. Only new ones. 

> Label changes flow with the PR comment timeline, so that would be good enough IMO.

Just wanted to point out the difference to the current behavior. I think if no one complains it's fine without. After all we trust the maintainers and the merge will trigger a notification for itself (and if someone complains adding the "do a new comment feature" should be easy).

**Comment by MartinNowak** [comment]
It's possible to fetch a merged commit created by github, they're available under refs/heads/ID/merge.
```d
git fetch upstream pull/6189/merge
git push upstream FETCH_HEAD:$base_branch
```
But the message is just `merge refA into refB`.

A problem w/ OAuth and labels it that they're lacking a connection. You don't notice a missing session in dlang-bot when adding a label on the PR page. Having to visit an extra site as w/ the auto-tester is a bit annoying OTOH.

**Comment by MartinNowak** [comment]
Much better idea:

We can integrate an auto-merge image button in the dlang-bot comment that links over to the dlang-bot site. Then we can just activate auto-merge, or do any necessary login&oauth stuff, then redirect back.

**Comment by wilzbach** [comment]
Okay it seems like we need this feature as auto-tester doesn't seem to like the required branches. I will have some time next Monday and can give it a go :)

**Comment by wilzbach** [comment]
A simple version based on special labels: https://github.com/MartinNowak/dlang-bot/pull/17

**Comment by wilzbach** [comment]
Solved with #17 :)
